providers/azure.go

```diff
@@ -151,6 +157,60 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (s
 		RefreshToken: jsonResponse.RefreshToken,
 	}
 	return
+
+}
+
+// RefreshSessionIfNeeded checks if the session has expired and uses the
+// RefreshToken to fetch a new ID token if required
+func (p *AzureProvider) RefreshSessionIfNeeded(ctx context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil || s.ExpiresOn.Add(-time.Minute*10).After(time.Now()) || s.RefreshToken == "" {
```

Comment:
Let's keep this aligned with the refresh timing logic in other providers:

```
	if s == nil || (s.ExpiresOn != nil && s.ExpiresOn.After(time.Now())) || s.RefreshToken == "" {
		return false, nil
	}
```

Refreshing sessions is an area we've identified corner case bugs around (refresh request stampedes getting multiple new access tokens, edge of time expiration/refresh token issues). We'd like to handle them better globally (I think to solve your issue, there's better logic that can surround how the `--cookie-refresh` timer logic in general if I recall).

Will this also lead to odd behavior for users that specific an access token TTL shorter than 10 minutes?

Reply:
Hmm yeah I assume this will do strange stuff when shorter TTLs are used. I'll remove the commit from this PR now and count on a global fix in the future. When I implemented this fix I was actually hunting down another bug, but I did not know that at that time and tried to fix it this way, so it should be ok to remove it again.